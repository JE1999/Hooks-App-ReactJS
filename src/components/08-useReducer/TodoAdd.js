import React, { memo } from 'react'
import { useForm } from '../../hooks/useForm'


export const TodoAdd = memo(({handleAddTodo}) => {

    const [ {desc}, handleChange, reset ] = useForm({desc: ''});

    const handleSubmit =  e =>{
        e.preventDefault()
        if(desc.trim().length <1){
            return
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: desc,
            done: false
        }

        handleAddTodo(newTodo)
        reset()
    }

    return (
        <>
            <h3>Agregar</h3>
                <form onSubmit={handleSubmit}>
                    <input 
                        type="text"
                        name="desc"
                        className="form-control"
                        placeholder="Agregar..."
                        autoComplete="off"
                        value={desc}
                        onChange={handleChange}
                    />
                    <button
                        type="submit"
                        className="btn btn-block btn-primary"
                    >
                        AGREGAR
                    </button>
                </form>
        </>
    )
})
