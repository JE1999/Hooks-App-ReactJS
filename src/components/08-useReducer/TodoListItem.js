import React, { memo } from 'react'

export const TodoListItem = memo(({todo, index, handleToggle, handleDelete}) => {
    return (
        <li 
            key={todo.id}
            className="list-group-item"
        >
            <p onClick={() => handleToggle(todo.id)} className={`${todo.done && 'text-success'}`} style={{cursor: 'pointer'}}>{index + 1}. {todo.desc}</p>
            <button 
                className="btn btn-sm btn-danger"
                onClick={() => handleDelete(todo.id)}
            >
                Borrar
            </button>
        </li>
    )
})

//Esto se incluye para saber cuantas veces este componente es llamado
//Test
//Solo se incluira si el componente esta rodeado por el memo
TodoListItem.displayName="TodoListItem"
