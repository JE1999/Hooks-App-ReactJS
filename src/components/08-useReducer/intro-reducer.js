const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = (state = initialState, action) => {

    if(action?.type === "AGREGAR"){
        return [...state, action.payload]
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
}

const action = {
    type: 'AGREGAR',
    payload: newTodo
}

todos = todoReducer(todos, action)

console.log(todos);