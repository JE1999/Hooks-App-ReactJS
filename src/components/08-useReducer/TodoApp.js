import React, { useReducer, useEffect, useCallback } from 'react'
import { todoReducer } from './todoReducer'
import { TodoList } from './TodoList'
import { TodoAdd } from './TodoAdd'

// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'App React',
//     done: false
// }];

const init = () =>{
    
    return JSON.parse(localStorage.getItem('todos')) || []
    
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'App React',
    //     done: false
    // }]
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    console.log(todos);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])

    const handleDelete = useCallback(todoId =>{
        const action = {
            type: "DELETE",
            payload: todoId
        }

        dispatch(action)
    },[])

    const handleToggle = useCallback(todoId =>{
        const action = {
            type: "TOGGLE",
            payload: todoId
        }

        dispatch(action)
    },[])

    const handleAddTodo = useCallback(newTodo => {
        dispatch({
            type: 'ADD',
            payload: newTodo
        })
    },[])

    return (
        <div className="container">
            <h1>TodoApp ({todos.length})</h1>
            <hr />

            <TodoAdd 
                handleAddTodo={handleAddTodo}
            />

            <hr />

            <TodoList
                todos={todos}
                handleToggle={handleToggle}
                handleDelete={handleDelete}
            />

        </div>
    )
}
