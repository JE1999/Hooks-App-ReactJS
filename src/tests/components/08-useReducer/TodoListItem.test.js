import React from 'react'
import { shallow } from 'enzyme'
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem'
import { demoTodos } from '../../fixtures/demoTodos'

describe('Pruebas de <TodoListItem />', () =>{

    const handleToggle = jest.fn()
    const handleDelete = jest.fn()

    const wrapper = shallow( 
        <TodoListItem 
            todo={demoTodos[0]}
            index={0}
            handleToggle={ handleToggle }
            handleDelete={ handleDelete }
        /> 
    )

    test('Mostrar componente', () => {
        
        expect(wrapper).toMatchSnapshot()
    
    })

    test('handleDelete();', () => {
        
        wrapper.find('button').simulate('click')
        expect( handleDelete ).toHaveBeenCalledWith( demoTodos[0].id )
    
    })

    test('handleToggle();', () => {
        
        wrapper.find('p').simulate('click')
        expect( handleToggle ).toHaveBeenCalledWith( demoTodos[0].id )
    
    })

    test('Mostrar text', () => {
        
        const p = wrapper.find('p')
        expect( p.text().trim() ).toBe( `1. ${demoTodos[0].desc}` )
    
    })

    test('Debe de tener la clase text-success si el todo.done = true', () => {
        
        const todo = demoTodos[0]
        todo.done = true

        const wrapper = shallow( 
            <TodoListItem 
                todo={ todo }
            /> 
        )

        expect( wrapper.find('p').hasClass('text-success') ).toBe(true)

    })

})