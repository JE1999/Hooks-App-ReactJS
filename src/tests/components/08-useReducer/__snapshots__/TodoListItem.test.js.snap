// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <TodoListItem /> Mostrar componente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    1
    . 
    Aprender ReactJS
  </p>
  <button
    className="btn btn-sm btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
